### PR TITLE
Add fix and test for orient-split roundtrip multiIndex

### PR DESCRIPTION
**Pull request by serhatgktp**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.0.0.rst` file if fixing a bug or adding a new feature.

`pandas.DataFrame.to_json()` and `pandas.read_json()` were not forming a successful round trip (i.e. a conversion from DataFrame to JSON, and then from JSON back to DataFrame) when used with a MultiIndex DataFrame with the setting orient="split".

This was fixed by adding a case for `orient="split"` for DataFrames with MultiIndex in `pandas.DataFrame.to_json()` that transposes the `columns` list in the output string.

Tests were added to `pandas/tests/io/json/test_pandas.py` under `test_roundtrip_multiindex()`.
The fix passes the added tests.

### CODE REVIEW

File: pandas/io/json/_json.py [hunk 1]
```diff
@@ -182,6 +182,20 @@ def to_json(
         indent=indent,
     ).write()
 
+    if orient == "split" and type(obj) == DataFrame \
+                            and type(obj.columns) == MultiIndex:
+        d = loads(s)
+        n = len(d["columns"])
+        new_cols_d = {i: [] for i in range(n)}
+        for i in range(n):
```
Comment (ma-anwar) [inline pandas/io/json/_json.py:190]: I understand why we're doing this; we want every first element of every list grouped together, every second element of every list grouped together etc.

I think there is a much cleaner way to do this that will be more readable using the zip method. I left an example below, could you try using this instead?

```
>>> for x in zip(['a', 'b', 'c'], ['d','e','f'], ['g', 'h', 'i']):
...     print(x)
...
('a', 'd', 'g')
('b', 'e', 'h')
('c', 'f', 'i')
```

File: pandas/io/json/_json.py [hunk 2]
```diff
@@ -182,6 +182,20 @@ def to_json(
         indent=indent,
     ).write()
 
+    if orient == "split" and type(obj) == DataFrame \
+                            and type(obj.columns) == MultiIndex:
+        d = loads(s)
```
Comment (ma-anwar) [inline pandas/io/json/_json.py:187]: I know the pandas code base uses one letter variables for things like strings and that's why you're using s below. But for our own variables that we're defining like `d`, can you give a more descriptive name. I understand here that we're converting the string to a dict, is there a better name you can use here so it's more obvious?

File: pandas/io/json/_json.py [hunk 3]
```diff
@@ -182,6 +182,20 @@ def to_json(
...
+        n = len(d["columns"])
+        new_cols_d = {i: [] for i in range(n)}
```
Comment (ma-anwar) [inline pandas/io/json/_json.py:189]: See my below comment on zip, once you incorporate that, can you see if this is necessary? I didn't give it much thought so it may be.

File: pandas/tests/io/json/test_pandas.py [hunk 4]
```diff
@@ -1843,3 +1843,13 @@ def test_json_uint64(self):
         df = DataFrame(data={"col1": [13342205958987758245, 12388075603347835679]})
         result = df.to_json(orient="split")
         assert result == expected
+
+    def test_roundtrip_multiindex(self):
```
Comment (ma-anwar) [inline pandas/tests/io/json/test_pandas.py:1847]: Thanks for being ahead of schedule and adding the test. Are there any edge cases to consider? Is an empty input possible by any chance? (i.e. orient split df that's empty?) Can you check if it's path of execution will go through your code and if so that the result is as expected?

**Review by ma-anwar — CHANGES_REQUESTED** [review]
Great work, I left some suggested changes. In general I understand the iterative approach you took but I think here a functional approach will be much cleaner for your problem. Can you give my suggestion a shot?

File: pandas/io/json/_json.py (hunk 1 above)
Comment (serhatgktp) [reply pandas/io/json/_json.py:190]: Thanks! I concur. Adding fixes now.

File: pandas/io/json/_json.py (hunk 2 above)
Comment (serhatgktp) [reply pandas/io/json/_json.py]: Done, thanks for bringing this up

File: pandas/tests/io/json/test_pandas.py (hunk 4 above)
Comment (serhatgktp) [reply pandas/tests/io/json/test_pandas.py:1847]: Good point, thanks! Adding a little intensity to the test here.

File: pandas/tests/io/json/test_pandas.py [hunk 5]
```diff
@@ -1843,3 +1843,36 @@ def test_json_uint64(self):
...
+                "multi_index": [[1, 2], [3, 4]],
+                "cols": [["2022", "2022"], ["JAN", "FEB"]]
+            },
+            "test_roundtrip_multiindex_three_levels": {
+                "multi_index": [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+                "cols": [["2022", "2022", "2022"], ["JAN", "FEB", "MAR"], ["MON", "TUE", "WED"]]
+            }
+        }
+
+        # Iterate over the test suite dictionary, unpacking the test name and test data
```
Comment (ma-anwar) [inline pandas/tests/io/json/test_pandas.py:1859]: Please remove this comment, your loop speaks for itself, nj with the clean code

File: pandas/tests/io/json/test_pandas.py [hunk 6]
```diff
@@ -1843,3 +1843,36 @@ def test_json_uint64(self):
...
         assert result == expected
+    
+    def test_roundtrip_multiindex(self):
+        test_suite = {
+            "test_roundtrip_multiindex": {
+                "multi_index": [[1, 2], [3, 4]],
+                "cols": [["2022", "2022"], ["JAN", "FEB"]]
+            },
+            "test_roundtrip_multiindex_three_levels": {
+                "multi_index": [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
```
Comment (ma-anwar) [inline pandas/tests/io/json/test_pandas.py:1854]: Is there any case where we can have empty values for any of these test parameters? Any possible edge cases?

Comment (serhatgktp) [reply pandas/tests/io/json/test_pandas.py:1854]: Our fix actually fails this test that I wrote:
```
	"test_roundtrip_multiindex_empty_3": {
		"multi_index": [[], []],
		"cols": [[], []]
	}
```

Here's the verbose output:
```
df1:
Empty DataFrame
Columns: []
Index: [0, 1]

json_str: {"columns":[],"index":[0,1],"data":[[],[]]}  # It looks like 'index' is 
                                                       # automatically set to [0,1]

df2:
Empty DataFrame
Columns: []
Index: [0, 1]
```

And here's the error message:

```
DataFrame.columns classes are not equivalent
[left]:  Float64Index([], dtype='float64')  # df2
[right]: MultiIndex([], )                   # df1
```

To me, it seems like the discrepancy is because the DataFrame generated by `read_json()` is not created as a DataFrame with MultiIndex since the JSON string simply has no columns. Thus it would be nonsensical for the result DataFrame to have MultiIndex columns. This is an edge case, but I don't think it poses a problem.

Comment (ma-anwar) [reply pandas/tests/io/json/test_pandas.py:1854]: Okay so just to confirm if you use this test, the path of execution does not go through the code that we're writing for this fix right?
